### PR TITLE
Allow global setting sound/music volume

### PR DESCRIPTION
**Pull request by pedro-w**
Also straightened up the interface to KMusic so both `set_volume` and `set_music_volume` take a single int param which is a volume 0 (silent) to 250 (loudest)

### CODE REVIEW

**Comment by OnlineCop** [comment]
Heh. I was working on something peripherally related: [dev/onlinecop/encapsulate_sounds](https://github.com/OnlineCop/kq-fork/tree/dev/onlinecop/encapsulate_sounds).

It was intended to move some of the sound/music initialization out of global variables into a common class, to help with future refactoring (mainly, to clean up the Configuration window and its logic).

**Comment by OnlineCop** [comment]
@pedro-w Will `music_mix.cpp` completely replace `music.cpp` as we're no longer using DUMB?

I'd vote to src/music_mix.cpp => src/music.cpp.

File: src/music_mix.cpp [hunk 1]
```diff
@@ -88,12 +88,16 @@ void KMusic::shutdown_music(void)
...
- * \param   volume 0 (silent) to 100 (loudest)
+ * \param   volume 0 (silent) to 250 (loudest)
  */
-void KMusic::set_music_volume(float volume)
+void KMusic::set_music_volume(int volume)
 {
-    int ivol = int(volume * float(MIX_MAX_VOLUME));
-    Mix_VolumeMusic(ivol);
+    float v = float(volume) / 250.0f;
+    if (v >= 0.0f && v <= 1.0f)
```
Comment (OnlineCop) [inline src/music_mix.cpp]: If the input volume is outside the valid range, would we want to do nothing, or just clamp the inputs?

`std::clamp()` from `<algorithm>` or `SDL_clamp()` from `SDL_stdinc.h` (probably included implicitly) could be used.

```c++
void KMusic::set_music_volume(int volume)
{
    constexpr float MaxMusicVolume = 250.0f; // or whatever will give a ratio within [0.0, 1.0]
    int ivol = std::clamp(int(float(MIX_MAX_VOLUME * volume) / MaxMusicVolume), 0, MIX_MAX_VOLUME);
    Mix_VolumeMusic(ivol);
}
```

I ran through some scenarios both with the old `[0,100]` range:
```
  inputs: -3, -2,  -1          ; outputs:   0,   0,   0
  inputs:  0,  1,   2,   3     ; outputs:   0,   1,   2,   3
  inputs: 98, 99, 100, 101, 102; outputs: 125, 126, 128, 128, 128
```
and with the new `[0,250]` range:
```
  inputs:  -3,  -2,  -1          ; outputs:   0,   0,   0
  inputs:   0,   1,   2,   3     ; outputs:   0,   0,   1,   1
  inputs:  98,  99, 100, 101, 102; outputs:  50,  50,  51,  51,  52
  inputs: 126, 127, 128, 129, 130; outputs:  64,  65,  65,  66,  66
  inputs: 248, 249, 250, 251, 252; outputs: 126, 127, 128, 128, 128
```

So all the clamping works as desired, if this is what we want. Some notes on the range change, though:

- When using the 0..100 range, the lower volumes like 1, 2 and 3 and upper volumes like 126, 127 and 128 increased the volume on each step (not a direct 1:1 correlation, but probably close enough).
- When using the 0..250 range, the input needs to be changed quite a bit more before we see changes, like seen above.

File: src/music_mix.cpp [hunk 2]
```diff
@@ -191,8 +195,18 @@ void KMusic::play_sample(void* chunk, int, int, int, int)
     }
     Mix_PlayChannel(-1, reinterpret_cast<Mix_Chunk*>(chunk), 0);
 }
-void KMusic::set_volume(float, int)
+/* \brief Set overall sound volume.
+ * \param   volume 0 (silent) to 250 (loudest)
+ */
+void KMusic::set_volume(int sound_volume)
```
Comment (OnlineCop) [inline src/music_mix.cpp]: Are music or sound volumes dependent on each other?

In other words, can I adjust the music volume without changing the sound volume, or adjust the sound volume without the music volume changing?

If this adjusts only sounds, could this be named `set_sound_volume`? If it adjusts both, could it be named something like `set_system_volume` or `set_overall_volume` or something?

File: src/setup.cpp [hunk 3]
```diff
@@ -69,12 +69,12 @@ static void parse_jb_setup(void);
 /*! \brief Play sound effects / music if adjusting it */
 static void sound_feedback(int val)
 {
-    Music.set_volume(val * 10, 0);
+    Music.set_volume(val * 10);
```
Comment (OnlineCop) [inline src/setup.cpp:72]: Do you know what this magic-number `10` represents?

Comment (OnlineCop) [reply src/setup.cpp:72]: I see a `10` on this line that we're dividing out:
```c++
    p = getavalue(_("Sound Volume"), 0, 25, gsvol / 10, true, sound_feedback);
```
I'm not familiar enough to see what it's trying to do, though.

Comment (pedro-w) [reply src/setup.cpp:72]: When you change the volume in the config screen it represents the volume with 25 blocks that you adjust with the arrow keys. The 10 is just for mapping blocks 0..25 to volume 0..250

File: src/music_mix.cpp (hunk 2 above)
Comment (pedro-w) [reply src/music_mix.cpp]: It was always supposed to - see 
https://github.com/OnlineCop/kq-fork/blob/8c638937d127b1bb705ea07682fdbb849fb32087/src/setup.cpp#L148
It means you can't have quiet sounds and loud music, would anyone want that?
In terms of the name, I'm not bothered. The code comment says it's 'overall sound volume' so I think it's clear.

File: src/music_mix.cpp (hunk 1 above)
Comment (pedro-w) [reply src/music_mix.cpp]: Looking at the old code I'm not sure there was ever a 0..100 range, despite the comment. The music volume seems to be passed in as a float 0.0..1.0, and it was derived by dividing by 250. e.g.
https://github.com/OnlineCop/kq-fork/blob/de7915f09b77bec9083d65334cff6793d0f70d63/src/setup.cpp#L444
It makes sense to me have 0..250 for both overall and music volumes when shown in the UI and saved to the config file. Then convert internally to whatever SDL / SDL_mixer require.
On clamping - possibly this was a holdover from Allegro - it had one command taking two params to set both digital audio and midi volumes, and you would pass -1 to say "don't change this one"
Clamp makes more sense.

**Comment by pedro-w** [comment]
I've made changes to use `std::clamp` - I think if you merge this PR before #87 it will go smoother

**Comment by pedro-w** [comment]
I also renamed the file, now if you merge #87 it will still show conflicts but much easier to resolve.

**Comment by OnlineCop** [comment]
Don't forget to change CMakeLists.txt from `src/music_mix.cpp` to `src/music.cpp`.

**Comment by z9484** [comment]
This crashes for me.
num_channels is 0 and line mixer.c:1072 is divide by 0

**Comment by OnlineCop** [comment]
I'm getting the same thing. I'm writing up a bug for this now.

**Comment by OnlineCop** [comment]
Created #89; it appears that starting a new game works fine but loading a saved game causes num_channels to be 0.

**Comment by pedro-w** [comment]
Seems like the music isn't getting initialized properly and we aren't checking it? It worked for me so may be only on Windows?

**Pull request merged**